runtime/common/src/lib.rs

```diff
@@ -301,76 +298,55 @@ pub mod investment_portfolios {
 	/// NOTE: Moving inner scope to any pallet would introduce tight(er)
 	/// coupling due to requirement of iterating over storage maps which in turn
 	/// require the pallet's Config trait.
-	pub fn get_account_portfolio<T, PoolInspector>(
-		investor: <T as frame_system::Config>::AccountId,
-	) -> Vec<(
-		<T as pallet_investments::Config>::InvestmentId,
-		InvestmentPortfolio<Balance, CurrencyId>,
-	)>
+	pub fn get_account_portfolio<T>(
+		investor: AccountId,
+	) -> Vec<(TrancheCurrency, InvestmentPortfolio<Balance, CurrencyId>)>
 	where
-		T: frame_system::Config
-			+ pallet_investments::Config
-			+ orml_tokens::Config
-			+ pallet_restricted_tokens::Config,
-		<T as pallet_investments::Config>::InvestmentId: TrancheCurrency<PoolId, TrancheId>
-			+ Into<<T as orml_tokens::Config>::CurrencyId>
-			+ Ord
-			+ Into<<T as pallet_restricted_tokens::Config>::CurrencyId>,
-		CurrencyId: From<<T as orml_tokens::Config>::CurrencyId>
-			+ From<<T as pallet_restricted_tokens::Config>::CurrencyId>,
-		<T as pallet_restricted_tokens::Config>::CurrencyId:
-			From<<T as orml_tokens::Config>::CurrencyId>,
-		Balance: From<<T as pallet_investments::Config>::Amount>
-			+ From<<T as pallet_restricted_tokens::Config>::Balance>,
-		PoolInspector: PoolInspect<
-			<T as frame_system::Config>::AccountId,
-			<T as pallet_restricted_tokens::Config>::CurrencyId,
-			PoolId = PoolId,
-			TrancheId = TrancheId,
-			Moment = Seconds,
-		>,
+		T: frame_system::Config<AccountId = AccountId>
+			+ pallet_investments::Config<InvestmentId = TrancheCurrency, Amount = Balance>
+			+ orml_tokens::Config<Balance = Balance, CurrencyId = CurrencyId>
+			+ pallet_restricted_tokens::Config<Balance = Balance, CurrencyId = CurrencyId>
+			+ pallet_pool_system::Config<PoolId = PoolId, CurrencyId = CurrencyId>,
 	{
-		let mut portfolio = BTreeMap::<
-			<T as pallet_investments::Config>::InvestmentId,
-			InvestmentPortfolio<Balance, CurrencyId>,
-		>::new();
+		let mut portfolio =
+			BTreeMap::<TrancheCurrency, InvestmentPortfolio<Balance, CurrencyId>>::new();
 
 		// Denote current tranche token balances before dry running collecting
 		orml_tokens::Accounts::<T>::iter_key_prefix(&investor).for_each(|currency| {
-			if let CurrencyId::Tranche(pool_id, tranche_id) = CurrencyId::from(currency) {
-				let pool_currency = PoolInspector::currency_for(pool_id)
-					.expect("Pool must exist; qed")
-					.into();
-				let free_balance = <pallet_restricted_tokens::Pallet<T> as fungibles::Inspect<
-					T::AccountId,
-				>>::reducible_balance(
-					currency.into(),
+			if let CurrencyId::Tranche(pool_id, tranche_id) = currency {
+				let pool_currency = pallet_pool_system::Pallet::<T>::currency_for(pool_id)
+					.expect("Pool must exist; qed");
```

Comment:
Even if it can no be broken right now, we should use `DispatchError::Other` in case future changes can break this. Changing it...

Reply:
Thanks for raising this question. I wasn't sure either so I tested this locally both with a local node and using Chopsticks for a temp testing runtime API which always panics. 

Luckily, it only creates a panic log but doesn't crash the client or of course not block production:

```
2024-06-07 11:20:11.778  INFO tokio-runtime-worker substrate: [🌀Parachain] 💤 Idle (0 peers), best: #25 (0x0992…4ffb), finalized #22 (0x2fcf…8920), ⬇ 99 B/s ⬆ 64 B/s
2024-06-07 11:20:12.050  INFO tokio-runtime-worker substrate: [Relaychain] ✨ Imported #70 (0x6065…02e4)

====================

Version: 0.10.39-22d0a3898a9

   0: backtrace::capture::Backtrace::create
   1: backtrace::capture::Backtrace::new
   2: sp_panic_handler::set::{{closure}}
   3: std::panicking::rust_panic_with_hook
   4: std::panicking::begin_panic_handler::{{closure}}
   5: std::sys_common::backtrace::__rust_end_short_backtrace
   6: _rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::option::expect_failed
   9: runtime_common::investment_portfolios::get_account_portfolio
  10: development_runtime::api::dispatch
  11: environmental::using
  12: sc_executor::executor::WasmExecutor<H>::with_instance::{{closure}}
  13: sc_executor::wasm_runtime::RuntimeCache::with_instance
  14: <sc_executor::executor::NativeElseWasmExecutor<D> as sp_core::traits::CodeExecutor>::call
  15: sp_state_machine::execution::StateMachine<B,H,Exec>::execute
  16: <sc_service::client::call_executor::LocalCallExecutor<Block,B,E> as sc_client_api::call_executor::CallExecutor<Block>>::call
  17: <sc_rpc::state::state_full::FullState<BE,Block,Client> as sc_rpc::state::StateBackend<Block,Client>>::call
  18: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
  19: tokio::runtime::task::core::Core<T,S>::poll
  20: tokio::runtime::task::harness::Harness<T,S>::poll
  21: tokio::runtime::blocking::pool::Inner::run
  22: std::sys_common::backtrace::__rust_begin_short_backtrace
  23: core::ops::function::FnOnce::call_once{{vtable.shim}}
  24: std::sys::pal::unix::thread::Thread::new::thread_start
  25: __pthread_joiner_wake


Thread 'tokio-runtime-worker' panicked at 'Pool must exist; qed', /Users/william/dev/rust/centrifuge-chain/runtime/common/src/lib.rs:314

This is a bug. Please report it at:

	https://github.com/paritytech/cumulus/issues/new

2024-06-07 11:20:12.151  WARN tokio-runtime-worker wasm-runtime: Evicting failed runtime instance error=Runtime panicked: Pool must exist; qed
2024-06-07 11:20:16.784  INFO tokio-runtime-worker substrate: [🌀Parachain] 💤 Idle (0 peers), best: #25 (0x0992…4ffb), finalized #23 (0x1500…5b5c), ⬇ 45 B/s ⬆ 45 B/s
2024-06-07 11:20:16.784  INFO tokio-runtime-worker substrate: [Relaychain] 💤 Idle (2 peers), best: #70 (0x6065…02e4), finalized #68 (0x311e…afe9), ⬇ 1.6kiB/s ⬆ 0.6kiB/s
2024-06-07 11:20:18.038  INFO tokio-runtime-worker babe: [Relaychain] 👶 New epoch 7 launching at block 0x3400…13fe (block slot 286292003 >= start slot 286292003).
2024-06-07 11:20:18.038  INFO tokio-runtime-worker babe: [Relaychain] 👶 Next epoch starts at slot 286292013
2024-06-07 11:20:18.043  INFO tokio-runtime-worker substrate: [Relaychain] ✨ Imported #71 (0x3400…13fe)
2024-06-07 11:20:18.045  INFO tokio-runtime-worker sc_basic_authorship::basic_authorship: [🌀Parachain] 🙌 Starting consensus session on top of parent 0x09920d2f8c5ce3ade2233a3252933f00beb5daac18c5c62a39320d9ca34e4ffb
2024-06-07 11:20:18.048  INFO tokio-runtime-worker sc_basic_authorship::basic_authorship: [🌀Parachain] 🎁 Prepared block for proposing at 26 (2 ms) [hash: 0x7efdb840f9b8be28758a156ce80ef05a0d8ce2bcef9a549de981143d8b864688; parent_hash: 0x0992…4ffb; extrinsics (2): [0x412f…fdf4, 0xd77a…184e]
2024-06-07 11:20:18.051  INFO tokio-runtime-worker aura::cumulus: [🌀Parachain] 🔖 Pre-sealed block for proposal at 26. Hash now 0xbd16575765e5d2053cb8f0164cc61921fad9778aa9a9f3055f2dc3c3d2a4c599, previously 0x7efdb840f9b8be28758a156ce80ef05a0d8ce2bcef9a549de981143d8b864688.
2024-06-07 11:20:18.051  INFO tokio-runtime-worker substrate: [🌀Parachain] ✨ Imported #26 (0xbd16…c599)
```

Anyways, if this code path here panics, we are in trouble anyways because it means we have a tranche currency in `orml_tokens` storage for a pool id which does not exist such that `currency_for(pool_id)` panics. We have many similar code paths outside of runtime API with the exact same assumptions of this being impossible.

Reply:
Fixed: [5baab79](https://github.com/centrifuge/centrifuge-chain/pull/1833/commits/5baab79af45dcce828672a12277e61a72dca0e60)

Reply:
Anyways, if this code path here panics, we are in trouble anyways because it means we have a tranche currency in `orml_tokens` storage for a pool id which does not exist such that `currency_for(pool_id)` panics. We have many similar code paths outside of runtime API with the exact same assumptions of this being impossible.

Reply:
Yeah, the problem with the non-compiled assumptions is that in the future they changed and silently... 💥

Reply:
Good test the above! I suppose substrate crate a tokio task to each requires. So if one panics nothing happens. TBH, that leaves me calm, maybe we can use more `expect()` for impossible things in some parts of our code. Do you want to revert the above fix?

Reply:
Although probably each panic would adds more memory leaks to the node 🤔

Reply:
I would error.